src/os_cloud_nodelet.cpp

```diff
@@ -152,31 +188,112 @@ class OusterCloud : public nodelet::Nodelet {
+            last_scan_last_nonzero_idx,
+            static_cast<double>(last_scan_last_nonzero_value),
+            scan_width + curr_scan_first_nonzero_idx,
+            static_cast<double>(curr_scan_first_nonzero_value), scan_width);
+        return ulround(interpolated_value);
+    }
+
+    uint64_t extrapolate_value(int curr_scan_first_nonzero_idx,
+                               uint64_t curr_scan_first_nonzero_value) {
+        double extrapolated_value =
+            curr_scan_first_nonzero_value -
+            scan_col_ts_spacing_ns * curr_scan_first_nonzero_idx;
+        return ulround(extrapolated_value);
+    }
+
+    // compute_scan_ts_0 for first scan
+    std::chrono::nanoseconds compute_scan_ts_0(
+        const ouster::LidarScan::Header<uint64_t>& ts_v) {
         auto idx = std::find_if(ts_v.data(), ts_v.data() + ts_v.size(),
                                 [](uint64_t h) { return h != 0; });
-        if (idx == ts_v.data() + ts_v.size()) return;
-        auto scan_ts = std::chrono::nanoseconds{ts_v(idx - ts_v.data())};
+        assert(idx != ts_v.data() + ts_v.size());  // should never happen
+        int curr_scan_first_nonzero_idx = idx - ts_v.data();
+        uint64_t curr_scan_first_nonzero_value = *idx;
+
+        uint64_t scan_ns =
+            curr_scan_first_nonzero_idx == 0
+                ? curr_scan_first_nonzero_value
+                : extrapolate_value(curr_scan_first_nonzero_idx,
+                                    curr_scan_first_nonzero_value);
+
+        last_scan_last_nonzero_idx =
+            find_if_reverse(ts_v, [](uint64_t h) { return h != 0; });
+        assert(last_scan_last_nonzero_idx >= 0);  // should never happen
+        last_scan_last_nonzero_value = ts_v(last_scan_last_nonzero_idx);
+        compute_scan_ts = [this](const auto& ts_v) {
+            return compute_scan_ts_n(ts_v);
+        };
+        return std::chrono::nanoseconds(scan_ns);
+    }
+
+    // compute_scan_ts_n applied to all subsequent scans except first one
+    std::chrono::nanoseconds compute_scan_ts_n(
+        const ouster::LidarScan::Header<uint64_t>& ts_v) {
+        auto idx = std::find_if(ts_v.data(), ts_v.data() + ts_v.size(),
+                                [](uint64_t h) { return h != 0; });
+        assert(idx != ts_v.data() + ts_v.size());  // should never happen
+        int curr_scan_first_nonzero_idx = idx - ts_v.data();
+        uint64_t curr_scan_first_nonzero_value = *idx;
+
+        uint64_t scan_ns = curr_scan_first_nonzero_idx == 0
+                               ? curr_scan_first_nonzero_value
+                               : impute_value(last_scan_last_nonzero_idx,
+                                              last_scan_last_nonzero_value,
+                                              curr_scan_first_nonzero_idx,
+                                              curr_scan_first_nonzero_value,
+                                              static_cast<int>(ts_v.size()));
```

Comment:
I don't handle this case but I think one could easily generalize the code to support it. Although, the value would be quite off but maybe better than nothing. In any case, I think we need to understand how often this happen and under what circumstances this take place.

Reply:
I just realized that we can use `ls.frame_id` for tracking such skips of one scan (not available for legacy profile, but all eUDPs give us frame_ids).

Btw in general if the flow of scans (i.e. lidar packets) interrupted, the next scan processed will have an incorrect scan_ts calculated because of the gap, maybe we need to add some bounds to the possible `scan_ts` results so it's always in some normal bounds from what can be with calculating it all using just a known sensor frequency and a first available column timestamp in LidarScan?

Reply:
Yes, `frame_id` does increase incrementally (on a good stream).

By interrupted means like sensor reset or reconnect? ROS1 driver doesn't handle reconnects.

Reply:
In the case of severe packet loss, it's possible to drop entire frames. Although at that point a lot of things are wrong... I could see the extrapolation using only the scan itself and `lidar_mode` being more useful than generalizing from the last seen scan (in some cases multiple frames ago) in that case

Reply:
The current implementation of **ScanBatcher** does not handle frame jumps or advances by more than one frame at a time. Anything that touches on this subject should be part of a larger effort to look into the current implementation of **ScanBatcher** itself. I think for a single sensor on a Gigabit network this isn't likely to happen, this maybe though of a concern when you have more than one sensor or an edge device with limited specs. That's why I would check which platforms or network setup that are affected by such severe packet drop, and do we need to handle this cases? Can we change the strategy of accumulating frames, expand the window of accepting packets from a previous frame, can 
we introduce a bit of delay before releasing the frame, etc. That's why I suggest any work in this regard should be part of a larger effort and is off scope for this ticket, including this thread.

Reply:
I think `ScanBatcher` does handle the frame jump since the last check is just `!=` but I think it's a fine decision to not handle it until we understand severe packet loss until we know more about what use cases we're really concerned with

Reply:
This does detect frame jumps but the handling isn't good I would consider it a proper handling of frame jumps. Imagine the case if you receive packets from two subsequent frames that were interleaved. Also imagine receiving a packet belonging to an older frame that was stuck in the network for some time. In both case which do happen BTW the client would see a sequence of frames with the frame_id jumping back and forth.